src/design_notes/general_coroutines.md

```diff
@@ -0,0 +1,258 @@
+- Thus, the "generator" syntax proposed in [eRFC-2033][1] and currently
+  implemented behind the "generator" feature is actually a coroutine syntax for
+  the sake of these notes, *not a true generator*.
+- Note also that "coroutines" here are really "semicoroutines" since they can
+  only yield back to their caller.
+- Generators are a specialization of coroutines.
+- I will continue to group the [original eRFC text][1] and the later [generator
+  resume arguments](https://github.com/rust-lang/rust/pull/68524) extension
+  togther as "[eRFC-2033][1]". That way I only have 3 big proposals to deal
+  with.
+
+```rust
+// This is a coroutine
+|stuff| {
+  yield stuff + 1;
+  yield stuff + 2;
+  return stuff + 3;
+}
+
+// This is a generator
+stream! {
+  yield 1;
+  yield 2;
+  yield 3;
+}
+```
+
+## Coroutine trait
+
+- The coroutine syntax must produce implementations of some trait.
+- [RFC-2781][2] and [eRFC-2033][1] propose the `Generator` trait.
+- Note that Rust's coroutines and subroutines look the same from the outside:
+  take input, mutate state, produce output.
+- Thus, [MCP-49][3] proposes using the `Fn*` traits instead, including a new
+  `FnPin` for immovable coroutines.
+  - Hierarchy: `Fn` is `FnMut + Unpin` is `FnPin` is `FnOnce`.
+    - May not be *required* at the trait level (someone may someday find a use
+      to implementing `FnMut + !FnPin`) but all closures implement the traits in
+      this order.
+
+## Coroutine syntax
+
+- The closure syntax is reused for coroutines by [eRFC-2033][1], [RFC-2781][2],
+  and [MCP-49][3].
+- Commentators have suggested that the differences between coroutines and
+  closures under [eRFC-2033][1] and [RFC-2781][2] justify an entirely distinct
+  syntax to reduce confusion.
+- [MCP-49][3] fully reuses the *semantics* of closures, greatly simplifying the
+  design space and making the shared syntax obvious.
+
+## Taking input
+
+- The major disagreement between past proposals is whether to use "yield
+  expressions" or "magic mutation".
+  - Yield expression: `let after = yield output;`
+    - Used by [eRFC-2033][1]
+  - Magic mutation: `let before = arg; yield output; let after = arg;`
+    - Used by [RFC-2781][2] and [MCP-49][3]
+- Many people have a strong gut preference for yield expressions.
+  - In simple cases, Rust generally prefers to produce values as output from
+    expressions rather than by mutation of state ⇒ yield expressions *feel* more
+    Rusty.
+  - I think it might be a bit like the `.await` bikeshed? People didn't *feel*
+    like `.await` was correct but it turned out to be well-justified.
+  - Justification ⇒ in addition to reasons below, holding references to past
+    resume args is rare, often a logic error. Rust can use mutation checks to
+    catch and give feedback.
+- "Magic mutation" is a bit of a misnomer. The resume argument values are *not
+  being mutated.* The argument bindings are simply being reassigned across
+  yields.
```

Comment:
The old values are (semantically speaking) overwritten, so the borrow checker will yell at the example you gave. The resume arg destructors run going into the `yield`, just like argument destructors are run going into a `return`.  The MCP explains that behavior extensively so I might add a link here back to it (even though it is closed?).

I worked for a while on a version of magic mutation where `yield` shadows old args instead of re-assigning them but it turned out to be a huge foot gun. I think it would be reasonable if I elaborate a bit on that in these notes.

Reply:
I'm a bit torn on the question of whether to link to MCPs for details of include them inline, but I'd be happy if you just want to link. I definitely tried to read this as a standalone document and maybe that was incorrect. I think ideally we'd "beef it up" so that it's the only thing one has to read, but that may include a section at the front like

### Read these first

* MCP xx, which explains the details of proposal 1
* RFC yy, which explains the details of proposal 2